fix(backend): check visibility of following/followers of remote users / feat: moderators can see following/followers of all users

tesaguri opened the pull request:
<!-- ℹ お読みください / README
PRありがとうございます！ PRを作成する前に、コントリビューションガイドをご確認ください:
Thank you for your PR! Before creating a PR, please check the contribution guide:
https://github.com/misskey-dev/misskey/blob/develop/CONTRIBUTING.md
-->

## What

リモートユーザのプロフィールを取得する際にActivityPubアクターの`following`と`followers`コレクションの取得を試み、その内容が取得できない場合は当該ユーザのフォロー・フォロワーの一覧を非公開としてデータベースに登録するようにします。

また、モデレータが全てのローカル・リモートユーザのフォロー・フォロワーの一覧を取得できるようにします。

## Why

Resolves #13362.

## Additional info (optional)

最低限の単体テストは追加していますが実際のサーバにおける連合の挙動は未確認です。

先行事例：mastodon/mastodon#11673（<https://github.com/mastodon/mastodon/blob/9d8dfeb5fbbc274482489a3ac9f22dd736da156c/app/services/activitypub/process_account_service.rb#L262-L281>）

本パッチでは「フォロワー限定」の公開範囲はとりあえず一律で非公開として扱うものとしています。

## Checklist
- [X] Read the [contribution guide](https://github.com/misskey-dev/misskey/blob/develop/CONTRIBUTING.md)
- [X] Test working in a local environment
- [ ] (If needed) Add story of storybook
- [X] (If needed) Update CHANGELOG.md
- [X] (If possible) Add tests

tesaguri committed:
fix(backend): check visibility of following/followers of remote users

Resolves https://github.com/misskey-dev/misskey/issues/13362.

tesaguri committed:
test(backend): add tests for visibility of following/followers of remote users

tesaguri committed:
docs(changelog): update CHANGELOG.md

Sayamame-beans commented:
13362をresolveするか微妙に思います
(要件に`※ただし https://github.com/misskey-dev/misskey/issues/13127 のようにモデレーションに影響が出ることが考えられる為権限がある場合は見えるように`があります。このPRではリモートユーザーのfollowing/follwersにprivateの概念を導入するものと思いますので、13362をcloseするには全ユーザーに対してこの要件を満たす実装を行ったPRを別途用意する必要がありそうです?)

tesaguri committed:
feat: moderators can see following/followers of all users

tesaguri commented:
#13128 に倣ってモデレータが非公開のfollowing/followersを見られるようにする機能もとりあえず本PRにまとめて追加しておきました。

tesaguri committed:
docs(changelog): update CHANGELOG.md

tesaguri commented:
（<!-- -->6d9161633bc1e6c86171aced67b10f02eb6dccf8<!-- -->と<!-- -->d9eeb8df9bd22e128f476288d3d4da036633f0db<!-- -->の間の変更は`CHANGELOG.md`のみであってバックエンドのE2Eテストが新たに落ちるような要素がないので、flakyな失敗とみなして良さそうですかね？）

Sayamame-beans commented:
(再試行で通りましたね)

tesaguri committed:
refactor(backend): minor refactoring

`createPerson`と`if`の条件を統一するとともに、異常系の
処理をearly returnに追い出すための変更。

zyoshoka commented:
モデレーターが `users/following` と `users/followers` を取得できるようにするのであれば、`users/show` で `followingCount` と `followersCount` を読めるようにするのが自然かと思います。

Related to #12214

tesaguri committed:
feat(backend): moderators can see following/followers count of all users

As per https://github.com/misskey-dev/misskey/pull/14375#issuecomment-2275044908.

syuilo commented:
🙏🏿🙏🏿

mei23 commented on packages/backend/src/core/activitypub/models/ApPersonService.ts:
```diff
@@ -703,4 +739,16 @@ export class ApPersonService implements OnModuleInit {
 	}
+
+	@bindThis
+	private async isPublicCollection(collection: string | ICollection | IOrderedCollection | undefined, resolver: Resolver): Promise<boolean> {
+		if (collection) {
+			const resolved = await resolver.resolveCollection(collection);
+			if (resolved.first || (resolved as ICollection).items || (resolved as IOrderedCollection).orderedItems) {
+				return true;
+			}
+		}
+
+		return false;
+	}
 }
```
itemsが0より上だった場合に公開扱いして良いのか？

Mastodonの~~すごくわかりづらい~~コードも、判定に使用しているのはfirstだけでitemsは考慮してないように見えるのだわね。
https://github.com/mastodon/mastodon/pull/11673/files#diff-0bc8e960429e6c65ff6b5f31c40c67ed4cab1904dcf2dcb9f13d1503338c84ef

tesaguri replied:
Activity Streams勧告の[Collections]節によると、Mastodonのように`first`プロパティで`CollectionPage`オブジェクトを指す形式の他に`items`を`Collection`に直接埋め込む形も規定されているので、単にMastodonのコードがMastodon以外のサーバを想定していないだけであろうと認識しています。

ちなみに`0`のような整数値を取るのは[`items`]でなく[`totalItems`]プロパティですね。

[Collections]: <https://www.w3.org/TR/2017/REC-activitystreams-core-20170523/#collections>
[`items`]: <https://www.w3.org/TR/2017/REC-activitystreams-vocabulary-20170523/#dfn-items>
[`totalItems`]: <https://www.w3.org/TR/2017/REC-activitystreams-vocabulary-20170523/#dfn-totalitems>

mei23 replied:
> [items](https://www.w3.org/TR/2017/REC-activitystreams-vocabulary-20170523/#dfn-items)でなく[totalItems](https://www.w3.org/TR/2017/REC-activitystreams-vocabulary-20170523/#dfn-totalitems)プロパティですね。

ごめんそれと間違ってたのだわ
